### DISCUSSION TITLE
How to pass custom property to the custom html report

### DISCUSSION THREAD

**Discussion by jeffreyrajanofficial** [detekt/detekt#7392] [Q&A]
I have created a CustomHtmlReport for my project and successfully generated the HTML report. Now, I want to pass a custom property to it. I need to generate the report based on this property. While I know how to pass property for custom rules in the config file, I'm not sure how to pass the property for the CustomHtmlReport since it's not added to the config file. Can anyone help me with passing the property to my CustomHtmlReport?

**Comment by cortinico** [comment]
> Now, I want to pass a custom property to it. I need to generate the report based on this property.

Can you clarify more what you're trying to achieve?

**Reply by jeffreyrajanofficial** [reply]
Below is my code 
```
class CustomHtmlReport: OutputReport() {
   override val ending = "html"

   override val render(detektion: Detektion) = 
        javaClass.getResource("/$DEFAULT_TEMPLATE")!!
            .openSafeStream()
            .bufferedReader()
            .use { it.readText() }
            .replace(PLACEHOLDER_VERSION, renderVersion())
            .replace(PLACEHOLDER_DATE, renderDate())
            .replace(PLACEHOLDER_METRICS, renderMetrics(detektion.metrics))
            .replace(PLACEHOLDER_COMPLEXITY_REPORT, renderComplexity(getComplexityMetrics(detektion)))
            .replace(PLACEHOLDER_ISSUES, renderIssues(detektion.issues))

 private renderMetrics(metrics: Collection<ProjectMetric>) = createHTML().div {
        // Writing some HTML logic 
    }
....
}
```

**app::build.gradle**

```
tasks.named("detekt").configure {
    reports {
        custom {
            reporteId = "CustomHtmlReport"
           outputLocation.set(file("build/reports/detekt/my_report")) 
        }
    }
}
```

**Command to run:** `./gradlew detekt --no-daemon`

This is the process I am following. I aim to create a metrics report based on some features. Therefore, I want to pass the feature name list either from build.gradle or config.yml. However, I am not sure how to pass the feature information to the CustomHtmlReport.

**Reply by cortinico** [reply]
> This is the process I am following. I aim to create a metrics report based on some features. Therefore, I want to pass the feature name list either from build.gradle or config.yml. However, I am not sure how to pass the feature information to the CustomHtmlReport.

It's still a bit unclear to me which kind of "some features" you're referring to so I don't fully understand if they should live inside the reporter or if it's something you want your users to configure.

Anyway, I would use Java Properties for this. You can add have a `resources/myconfig.properties` file that you can access natively from Java/Kotlin. Here an example of how to do it: https://stackoverflow.com/a/15749281

**Reply by jeffreyrajanofficial** [reply]
I referred to the Home screen, Timeline screen, etc., as features.

Thanks, I will try the Java Properties

**Comment by BraisGabin** [comment]
On the report you receive a Detektion instance. That class allows you to add arbitrary values to it and then you can read them in the custom reporter.

**Reply by jeffreyrajanofficial** [reply]
I have created CustomDetektModule. I will be using it in different Modules (HomeModules, TimelineModule). Hence the Module Owner will decide which path is considered as a feature. So I will have to make this as a dynamic path which can be passed from different module. I believe the Detektion instance won't be able to pass the path from each module (please correct me if I am wrong).

Note: I am referring **path** as a custom parameter which I will be using to pass the value to the CustomDetektModule::CustomReport

**Reply by BraisGabin** [reply]
Detektion implements [UserDataHolder](https://github.com/JetBrains/intellij-community/blob/master/platform/util/src/com/intellij/openapi/util/UserDataHolder.java). With that you can store whatever you want. Maybe I'm missing something.
